packages/flutter_tools/lib/src/version.dart

```diff
@@ -1263,3 +1300,70 @@ const String _newVersionAvailableMessage = '''
 A new version of Flutter is available!
 
 To update to the latest version, run "flutter upgrade".''';
+
+/// Data class for parsing engine_stamp.json.
+@visibleForTesting
+final class FlutterEngineStampFromFile {
+  FlutterEngineStampFromFile._({
+    required this.buildDate,
+    required this.gitRevision,
+    required this.gitRevisionDate,
+    required this.contentHash,
+  });
+
+  /// General time when the engine artifacts were built.
+  final DateTime buildDate;
+
+  /// The git commit sha for these engine artifacts.
+  final String gitRevision;
+
+  /// The date of the [gitRevision] commit.
+  final DateTime gitRevisionDate;
+
+  /// The content-aware hash for this build.
+  ///
+  /// This hash helps uniquely identify engine changes and reduce artifact builds.
+  final String contentHash;
+
+  /// Attempt to parse [jsonFile] for `engine_stamp.json` values.
+  static FlutterEngineStampFromFile? tryParseFromFile(File jsonFile) {
+    final Map<String, Object?> data;
+    try {
+      data = json.decode(jsonFile.readAsStringSync()) as Map<String, Object?>;
+
+      if (data case {
+        'build_time_ms': final int buildTimeMs,
+        'git_revision': final String gitRevision,
+        'git_revision_date': final String gitRevisionDate,
+        'content_hash': final String contentHash,
+      }) {
+        return FlutterEngineStampFromFile._(
+          buildDate: DateTime.fromMillisecondsSinceEpoch(buildTimeMs),
+          gitRevision: gitRevision,
+          gitRevisionDate: DateTime.parse(gitRevisionDate),
+          contentHash: contentHash,
+        );
+      }
+      // ignore: avoid_catches_without_on_clauses
+    } catch (_) {
+      return null;
+    }
+    return null;
+  }
+}
+
+/// Simple "X (months|days|hours|minutes) ago" converter.
+extension on Duration {
```

Comment:
This seems generally useful. Consider moving it into `lib/src/base/utils.dart` so it can be more discoverable.

Reply:
Will do on Monday.

Reply:
Moved